File: EmployeeShifts.py
# An application that manages employee shifts in a company.
# Shrisan Kapali
# Student ID: 005032249
# Assignment 4 : Implementing Control Structures

import random
import tkinter as tk
from tkinter import ttk
from collections import defaultdict

# GLobal constants for the application
# Days of the week and shifts
DAYS_OF_WEEK = [
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
    "Sunday",
]

SHIFTS = ["Morning", "Afternoon", "Evening"]


# A method that creates a weekly schedule based on employee preferences.
def create_schedule(
    employee_preferences, num_employees_per_shift=2, max_work_days_per_week=5
):

    # Initialize the weekly schedule structure
    weekly_schedule = {day: {shift: [] for shift in SHIFTS} for day in DAYS_OF_WEEK}

    # Initialize a counter for days worked by each employee
    employee_work_days = defaultdict(int)
    for emp_name in employee_preferences:
        employee_work_days[emp_name] = 0

    unassigned_employees = []  # To track employees who couldn't be fully scheduled

    print("**********************************************************")
    print("************* Starting Schedule Generation ***************")
    print("**********************************************************")

    # Assigning employees based on their day preferences ---
    # Looping through each day of the week
    for day in DAYS_OF_WEEK:
        print(f"\nCurrently scheduling for {day}:")
        # Need to make sure that each day has at least 2 employees per shift
        # A tracker that helps identify if any employees were assigned for this day
        employees_assigned_today = defaultdict(bool)

        # Bonus implementation: Assigned employees based on their preferences
        # Looping through each employee's preferences
        for employee_name, preferences in employee_preferences.items():
            # As each employee cannot work more than 5 days per week
            if employee_work_days[employee_name] >= max_work_days_per_week:
                print(
                    f"  {employee_name} has reached max work days ({max_work_days_per_week})."
                )
                continue

            # As employee can only work 1 shift per day, we check if they are already assigned today
            # If they are already assigned for this day, skip to the next employee
            if employees_assigned_today[employee_name]:
                continue

            # Get the preferred shifts for the current day from employee preferences
            preferred_shifts_on_day = preferences.get(day, [])

            # If the employee has no preferences for this day, they are unassigned
            assigned_this_day = False
            for preferred_shift in preferred_shifts_on_day:
                # Validate shift name
                if preferred_shift not in SHIFTS:
                    print(
                        f"    Warning: {employee_name} does not prefer working on Shift '{preferred_shift}' for  {day}. Skipping to next."
                    )
                    continue

                # Only if the preferred shift is not full, assign the employee
                if (
                    len(weekly_schedule[day][preferred_shift])
                    < num_employees_per_shift + 1
                ):
                    weekly_schedule[day][preferred_shift].append(employee_name)
                    employees_assigned_today[employee_name] = True
                    employee_work_days[employee_name] += 1
                    print(
                        f"    Assigned {employee_name} to preferred {preferred_shift} on {day}."
                    )
                    assigned_this_day = True
                    break  # Employee assigned for this day, move to next employee
                else:
                    # Preferred shift is full, try next preference if available
                    print(
                        f"    {employee_name}'s preferred {preferred_shift} on {day} is full. Trying next preference."
                    )
                    pass  # Try next preference in the list

            if not assigned_this_day and not employees_assigned_today[employee_name]:
                # If employee couldn't be assigned based on preferences for this day,
                # they are still available for random assignment
                print(
                    f"    {employee_name} could not be assigned based on preferences for {day}."
                )
                pass  # Will be handled in the lower for loop code if needed

        # --- Fill remaining shifts to meet minimums (2 employees per shift) ---
        # And also catch employees who couldn't be assigned by preference but are available
        for shift_type in SHIFTS:
            while len(weekly_schedule[day][shift_type]) < num_employees_per_shift:
                # Find available employees who haven't worked max days and are not assigned today
                available_for_random_assignment = []
                for emp_name in employee_preferences.keys():
                    if (
                        employee_work_days[emp_name] < max_work_days_per_week
                        and not employees_assigned_today[emp_name]
                    ):
                        available_for_random_assignment.append(emp_name)

                if available_for_random_assignment:
                    # Randomly pick an employee from the available pool
                    chosen_employee = random.choice(available_for_random_assignment)
                    weekly_schedule[day][shift_type].append(chosen_employee)
                    employees_assigned_today[chosen_employee] = True
                    employee_work_days[chosen_employee] += 1
                    print(
                        f"    Randomly assigned {chosen_employee} to {shift_type} on {day} to meet minimum."
                    )
                else:
                    print(
                        f"    WARNING: Could not meet minimum of {num_employees_per_shift} for {shift_type} on {day}. Current: {len(weekly_schedule[day][shift_type])}"
                    )
                    break  # Cannot find more employees for this shift

    # For not working Employees
    for emp_name, days_worked in employee_work_days.items():
        if days_worked == 0:
            unassigned_employees.append(emp_name)

    print("\n--- Schedule Generation Complete ---")

    return weekly_schedule, employee_work_days, unassigned_employees


# A method to print the generated schedule in a readable format.
def display_schedule(schedule, employee_work_days):
    print("\n" + "=" * 50)
    print("                 FINAL WEEKLY SCHEDULE")
    print("=" * 50)

    for day in DAYS_OF_WEEK:
        print(f"\n--- {day.upper()} ---")
        for shift_type in SHIFTS:
            employees = schedule[day][shift_type]
            if employees:
                print(f"  {shift_type:<10}: {', '.join(employees)}")
            else:
                print(f"  {shift_type:<10}: No employees assigned")

    print("\n" + "=" * 50)
    print("                 EMPLOYEE WORK DAYS SUMMARY")
    print("=" * 50)
    for employee, days in sorted(employee_work_days.items()):
        print(f"{employee:<15}: {days} days worked")
    print("=" * 50)


# Showing the schedule in a colorful calendar-style Tkinter GUI
def display_schedule_ui(schedule):
    SHIFT_COLORS = {
        "Morning": "#ffffff",
        "Afternoon": "#ffffff",
        "Evening": "#ffffff",
    }

    HEADER_BG = "#2196f3"  # Dark blue-gray
    HEADER_FG = "white"
    SHIFT_HEADER_BG = "#2196f3"
    SHIFT_HEADER_FG = "white"

    root = tk.Tk()
    root.title("Weekly Employee Schedule")

    style = ttk.Style()
    style.configure("TLabel", font=("Segoe UI", 10), padding=5)

    # Create main frame
    main_frame = ttk.Frame(root, padding=10)
    main_frame.grid(row=0, column=0, sticky="NSEW")

    # Column headers: Days
    for col, day in enumerate([""] + DAYS_OF_WEEK):
        bg = HEADER_BG if col != 0 else SHIFT_HEADER_BG
        fg = HEADER_FG if col != 0 else SHIFT_HEADER_FG
        header = tk.Label(
            main_frame,
            text=day,
            font=("Segoe UI", 10, "bold"),
            bg=bg,
            fg=fg,
            padx=10,
            pady=5,
            relief="ridge",
            bd=2,
        )
        header.grid(row=0, column=col, sticky="NSEW")

    # Shift rows and schedule cells
    for row, shift in enumerate(SHIFTS, start=1):
        # Shift label (left column)
        tk.Label(
            main_frame,
            text=shift,
            font=("Segoe UI", 10, "bold"),
            bg=SHIFT_HEADER_BG,
            fg=SHIFT_HEADER_FG,
            padx=10,
            pady=5,
            relief="ridge",
            bd=2,
        ).grid(row=row, column=0, sticky="NSEW")

        for col, day in enumerate(DAYS_OF_WEEK, start=1):
            employees = schedule[day][shift]
            bg_color = SHIFT_COLORS.get(shift, "#ffffff")
            cell_text = "\n".join(employees) if employees else "No one assigned"
            fg_color = "#000000" if employees else "#888888"

            # Create schedule cell
            tk.Label(
                main_frame,
                text=cell_text,
                bg=bg_color,
                fg=fg_color,
                font=("Segoe UI", 9),
                padx=5,
                pady=5,
                justify="center",
                relief="solid",
                bd=1,
                wraplength=120,
            ).grid(row=row, column=col, sticky="NSEW")

    # Grid expand/resizing
    for i in range(len(DAYS_OF_WEEK) + 1):
        main_frame.columnconfigure(i, weight=1)
    for i in range(len(SHIFTS) + 1):
        main_frame.rowconfigure(i, weight=1)

    root.mainloop()


# Display a summary of employee work days in a separate UI window
def display_employee_summary_ui(employee_work_days):
    root = tk.Tk()
    root.title("Employee Work Summary")

    header_bg = "#1e88e5"
    header_fg = "white"
    row_bg = "#ffffff"

    summary_frame = ttk.Frame(root, padding=10)
    summary_frame.grid(row=0, column=0, sticky="NSEW")

    # Header labels
    tk.Label(
        summary_frame,
        text="Employee",
        font=("Segoe UI", 10, "bold"),
        bg=header_bg,
        fg=header_fg,
        padx=10,
        pady=5,
        relief="ridge",
        bd=2,
    ).grid(row=0, column=0, sticky="NSEW")

    tk.Label(
        summary_frame,
        text="Shifts Assigned",
        font=("Segoe UI", 10, "bold"),
        bg=header_bg,
        fg=header_fg,
        padx=10,
        pady=5,
        relief="ridge",
        bd=2,
    ).grid(row=0, column=1, sticky="NSEW")

    # Employee data rows
    for row, (emp, count) in enumerate(sorted(employee_work_days.items()), start=1):
        tk.Label(
            summary_frame,
            text=emp,
            font=("Segoe UI", 10),
            bg=row_bg,
            padx=10,
            pady=5,
            relief="solid",
            bd=1,
        ).grid(row=row, column=0, sticky="NSEW")

        tk.Label(
            summary_frame,
            text=str(count),
            font=("Segoe UI", 10),
            bg=row_bg,
            padx=10,
            pady=5,
            relief="solid",
            bd=1,
        ).grid(row=row, column=1, sticky="NSEW")

    # Grid expansion
    summary_frame.columnconfigure(0, weight=1)
    summary_frame.columnconfigure(1, weight=1)

    root.mainloop()


# --- Example Usage ---
if __name__ == "__main__":
    # Each employee has preferences for each day. If a day is not listed, they have no preference.
    # Preferences are lists, indicating priority.
    sample_employee_preferences = {
        "Ricky": {
            "Monday": ["Afternoon", "Evening"],
            "Tuesday": ["Evening", "Morning"],
            "Thursday": ["Morning"],
            "Friday": ["Morning"],
        },
        "John": {
            "Tuesday": ["Morning"],
            "Wednesday": ["Morning"],
            "Thursday": ["Morning"],
            "Friday": ["Morning"],
            "Saturday": ["Morning"],
            "Sunday": ["Morning"],
        },
        "Nuguid": {
            "Monday": ["Evening"],
            "Tuesday": ["Evening"],
            "Wednesday": ["Evening"],
            "Thursday": ["Evening"],
            "Friday": ["Evening"],
            "Saturday": ["Evening"],
        },
        "Elyza": {
            "Monday": ["Evening", "Morning"],
            "Tuesday": ["Morning"],
            "Wednesday": ["Evening"],
            "Thursday": ["Afternoon"],
            "Friday": ["Morning"],
            "Saturday": ["Afternoon"],
            "Sunday": ["Morning"],
        },
        "Allison": {
            "Monday": ["Morning", "Evening"],
            "Tuesday": ["Afternoon"],
            "Wednesday": ["Morning"],
            "Thursday": ["Afternoon"],
            "Friday": ["Evening"],
            "Saturday": ["Evening"],
            "Sunday": ["Afternoon"],
        },
        "Henry": {
            "Monday": ["Afternoon"],
            "Tuesday": ["Evening"],
            "Wednesday": ["Afternoon"],
            "Thursday": ["Evening"],
            "Friday": ["Morning"],
            "Saturday": ["Morning"],
            "Sunday": ["Evening"],
        },
        "Cavil": {
            "Monday": ["Evening"],
            "Tuesday": ["Morning"],
            "Wednesday": ["Morning"],
            "Thursday": ["Morning", "Evening"],
            "Friday": ["Afternoon", "Evening"],
            "Saturday": ["Afternoon", "Evening"],
            "Sunday": ["Morning", "Evening"],
        },
        # No preferences at all
        "Merit": {},
        "Braxton": {},
        "Neil": {},
        "Marko": {},
    }

    # 2. Scheduling Logic (called by create_schedule function)
    # 3. Shift Conflicts (handled within create_schedule, prioritizing same-day alternatives
    #    based on preference order and then random assignment for minimums)
    final_schedule, employee_days, unassigned = create_schedule(
        sample_employee_preferences
    )

    # 4. Output: Display the final schedule
    display_schedule(final_schedule, employee_days)

    # Graphical Display of schdule and the number of shifts assigned to each employee
    display_schedule_ui(final_schedule)
    display_employee_summary_ui(employee_days)

    # Output any employees who couldn't be assigned at all
    if unassigned:
        print(
            "\nNote: The following employees were not assigned for some days due to conflicts or limits:"
        )
        for emp in unassigned:
            print(f"- {emp}")

File: EmployeeShifts.java
import java.util.*;

public class EmployeeShifts {
    // Constants for scheduling rules
    static final int MAX_DAYS = 5;
    static final int MIN_EMPLOYEES = 2;

    // Enum for days of the week
    enum DayOfWeek {
        SUNDAY, MONDAY, TUESDAY, WEDNESDAY, THURSDAY, FRIDAY, SATURDAY
    }

    // Enum for shift types
    enum ShiftType {
        MORNING, AFTERNOON, EVENING
    }

    // Static Employee class to represent each employee
    static class Employee {
        // Name of the employee
        String name;

        // Store Day preferences for Employee
        Map<DayOfWeek, ShiftType> preferredShifts;

        // How many days employee has worked this week
        int daysWorkedThisWeek;

        // Helper to flag if employee has a shift today
        boolean hasShiftToday;

        // Contructor to initialize an employee with a name and empty preferences
        public Employee(String name) {
            this.name = name;
            this.preferredShifts = new HashMap<>();
            this.daysWorkedThisWeek = 0;
            this.hasShiftToday = false;
        }

        // Getters for employee properties
        public String getName() {
            return name;
        }

        public Map<DayOfWeek, ShiftType> getPreferredShifts() {
            return preferredShifts;
        }

        public int getDaysWorkedThisWeek() {
            return daysWorkedThisWeek;
        }

        public void incrementDaysWorked() {
            this.daysWorkedThisWeek++;
        }

        public boolean hasShiftToday() {
            return hasShiftToday;
        }

        public void setHasShiftToday(boolean hasShiftToday) {
            this.hasShiftToday = hasShiftToday;
        }

        @Override
        public String toString() {
            return name;
        }
    }

    // Weekly company schedule
    static Map<DayOfWeek, Map<ShiftType, List<Employee>>> weeklySchedule;

    // A list to store all registered employees
    static List<Employee> employees;

    public static void main(String[] args) {
        // Initialize data structures
        employees = new ArrayList<>();
        weeklySchedule = new EnumMap<>(DayOfWeek.class);

        // Initialize weekly schedule with empty lists for each day and shift
        for (DayOfWeek day : DayOfWeek.values()) {
            weeklySchedule.put(day, new EnumMap<>(ShiftType.class));
            for (ShiftType shift : ShiftType.values()) {
                weeklySchedule.get(day).put(shift, new ArrayList<>());
            }
        }

        // Test Employee Data
        System.out.println("--- Employees and their schedule preferences ---");
        // Calling Helper class to load employees
        // Switch between loading employees with or without preferences
        // loadEmployeesWithPreferences(employees);
        loadEmployeesWithoutPreferences(employees);

        System.out.println("\n--- Employee has been loaded  ---");
        System.out.println("\n--- Now Generating Schedule ---");

        // Looping through each day of the week to assign shifts
        for (DayOfWeek day : DayOfWeek.values()) {
            // Initailly resetting the entire employee's shift status for the day to false
            for (Employee emp : employees) {
                emp.setHasShiftToday(false);
            }

            // Begin scheduling for the current day
            System.out.println("\nScheduling Shifts for " + day + ":");

            // Based on employee preferences, assign shifts
            for (Employee emp : employees) {
                // If the employee has worked less than 5 days this week and doesn't have a
                // shift
                if (emp.getDaysWorkedThisWeek() < MAX_DAYS && !emp.hasShiftToday()) {
                    // Does the employee have a preferred shift for this day?
                    ShiftType preferredShift = emp.getPreferredShifts().get(day);
                    // If they do, try to assign them to that shift
                    if (preferredShift != null) {
                        List<Employee> shiftEmployees = weeklySchedule.get(day).get(preferredShift);

                        // Initially load 2 people per each shift to fill all the calendar
                        if (shiftEmployees.size() < MIN_EMPLOYEES) {
                            shiftEmployees.add(emp);
                            emp.incrementDaysWorked();
                            emp.setHasShiftToday(true);
                            System.out.println(
                                    emp.getName() + " has been assigned to " + preferredShift + " shift.");
                        } else {
                            System.out.println(emp.getName() + "'s preferred " + preferredShift
                                    + " has been filled for " + day + ".");
                            // Check if the employee can be assigned to another shift
                            assignAlternateShift(emp, day);
                        }
                    }
                    // If the employee has worked 5 days, skip them
                } else if (emp.getDaysWorkedThisWeek() >= MAX_DAYS) {
                    System.out.println(emp.getName() + " has worked for the maximum days this week ("
                            + emp.getDaysWorkedThisWeek() + ").");
                }
            }

            Random random = new Random();

            // If the preferred shifts are not filled, try to fill them with unassigned
            // employees
            for (ShiftType shift : ShiftType.values()) {
                // Get the current list of employees assigned to this shift
                List<Employee> currentShiftEmployees = weeklySchedule.get(day).get(shift);

                // Making sure each shift has at least 2 employees
                while (currentShiftEmployees.size() < MIN_EMPLOYEES) {
                    System.out.println(shift + " shift for " + day + " has not met the minimum employee count ("
                            + currentShiftEmployees.size() + ").");
                    // Finding employees who have not been assigned today and have worked less than
                    // 5 days
                    Employee assignedFiller = null;
                    List<Employee> availableFillers = new ArrayList<>();
                    for (Employee emp : employees) {
                        if (emp.getDaysWorkedThisWeek() < MAX_DAYS && !emp.hasShiftToday()) {
                            availableFillers.add(emp);
                        }
                    }

                    // If there are available employees to fill the shift
                    if (!availableFillers.isEmpty()) {
                        // Randomly select an employee from the available fillers
                        assignedFiller = availableFillers.get(random.nextInt(availableFillers.size()));
                        currentShiftEmployees.add(assignedFiller);
                        assignedFiller.incrementDaysWorked();
                        assignedFiller.setHasShiftToday(true);
                        System.out.println(
                                assignedFiller.getName() + " has been assigned to " + shift + " shift.");
                    } else {
                        System.out.println(
                                "No additional employees is available to fill " + shift + " shift for " + day
                                        + ".");
                        break;
                    }
                }
            }
        }

        // Now that the schedule is generated, print the final weekly schedule
        System.out.println("\n--- Weekly Employee Shift Schedule ---");

        // Loop through each day of the week and print the assigned employees for each
        for (DayOfWeek day : DayOfWeek.values()) {
            System.out.println("\n" + day + ":");
            for (ShiftType shift : ShiftType.values()) {
                List<Employee> assignedEmployees = weeklySchedule.get(day).get(shift);
                System.out.print("  " + shift + ": ");
                // If the list is empty, no employees assigned
                if (assignedEmployees.isEmpty()) {
                    System.out.println("No employees assigned.");
                } else {
                    // Print he names of assigned employees
                    for (int i = 0; i < assignedEmployees.size(); i++) {
                        System.out.print(assignedEmployees.get(i).getName());
                        if (i < assignedEmployees.size() - 1) {
                            System.out.print(", ");
                        }
                    }
                    System.out.println("(Total: " + assignedEmployees.size() + ")");
                }
            }
        }

        System.out.println("\n--- Employee total shifts summary ---");
        for (Employee emp : employees) {
            System.out.println(emp.getName() + " worked " + emp.getDaysWorkedThisWeek() + " days.");
        }
    }

    // Sheduling employee for alternate shifts if their preferred shift is full
    private static void assignAlternateShift(Employee emp, DayOfWeek day) {
        ShiftType[] shiftsToTry = ShiftType.values();

        for (ShiftType altShift : shiftsToTry) {
            // Skip preferred shift
            if (emp.getPreferredShifts().get(day) != null && emp.getPreferredShifts().get(day) == altShift) {
                continue;
            }

            List<Employee> altShiftEmployees = weeklySchedule.get(day).get(altShift);

            // Only if the alternate shift has less than the minimum employees
            if (altShiftEmployees.size() < MIN_EMPLOYEES && !emp.hasShiftToday()) {
                altShiftEmployees.add(emp);
                emp.incrementDaysWorked();
                emp.setHasShiftToday(true);
                System.out.println("Employee : " + emp.getName() + " assigned to " + altShift + " shift on "
                        + day + ".");
                return;
            }
        }

        // If no alternate shift could be assigned
        System.out.println(emp.getName() + " cannot be assigned to alternate shift on " + day
                + " (all full or already assigned).");
    }

    // Helper Class to load employees
    static void loadEmployeesWithoutPreferences(List<Employee> employees) {
        employees.add(new Employee("Mary"));
        employees.add(new Employee("Kate"));
        employees.add(new Employee("Kevin"));
        employees.add(new Employee("Patrick"));
        employees.add(new Employee("Evan"));
        employees.add(new Employee("Joe"));
    }

    static void loadEmployeesWithPreferences(List<Employee> employees) {
        // Employee List Initialization
        Employee emp1 = new Employee("Mary");
        emp1.getPreferredShifts().put(DayOfWeek.MONDAY, ShiftType.MORNING);
        emp1.getPreferredShifts().put(DayOfWeek.TUESDAY, ShiftType.MORNING);
        emp1.getPreferredShifts().put(DayOfWeek.WEDNESDAY, ShiftType.MORNING);
        employees.add(emp1);

        Employee emp2 = new Employee("Kate");
        emp2.getPreferredShifts().put(DayOfWeek.WEDNESDAY, ShiftType.EVENING);
        emp2.getPreferredShifts().put(DayOfWeek.FRIDAY, ShiftType.EVENING);
        emp2.getPreferredShifts().put(DayOfWeek.SATURDAY, ShiftType.EVENING);
        employees.add(emp2);

        Employee emp3 = new Employee("Kevin");
        emp3.getPreferredShifts().put(DayOfWeek.TUESDAY, ShiftType.AFTERNOON);
        emp3.getPreferredShifts().put(DayOfWeek.WEDNESDAY, ShiftType.AFTERNOON);
        emp3.getPreferredShifts().put(DayOfWeek.THURSDAY, ShiftType.AFTERNOON);
        emp3.getPreferredShifts().put(DayOfWeek.SATURDAY, ShiftType.AFTERNOON);
        emp3.getPreferredShifts().put(DayOfWeek.SUNDAY, ShiftType.AFTERNOON);
        employees.add(emp3);

        Employee emp4 = new Employee("Patrick");
        emp4.getPreferredShifts().put(DayOfWeek.MONDAY, ShiftType.MORNING);
        emp4.getPreferredShifts().put(DayOfWeek.TUESDAY, ShiftType.MORNING);
        emp4.getPreferredShifts().put(DayOfWeek.WEDNESDAY, ShiftType.MORNING);
        emp4.getPreferredShifts().put(DayOfWeek.THURSDAY, ShiftType.MORNING);
        emp4.getPreferredShifts().put(DayOfWeek.FRIDAY, ShiftType.MORNING);
        emp4.getPreferredShifts().put(DayOfWeek.SATURDAY, ShiftType.MORNING);
        emp4.getPreferredShifts().put(DayOfWeek.SUNDAY, ShiftType.MORNING);
        employees.add(emp4);

        Employee emp5 = new Employee("Evan");
        emp5.getPreferredShifts().put(DayOfWeek.MONDAY, ShiftType.EVENING);
        emp5.getPreferredShifts().put(DayOfWeek.TUESDAY, ShiftType.MORNING);
        emp5.getPreferredShifts().put(DayOfWeek.WEDNESDAY, ShiftType.EVENING);
        emp5.getPreferredShifts().put(DayOfWeek.THURSDAY, ShiftType.AFTERNOON);
        emp5.getPreferredShifts().put(DayOfWeek.FRIDAY, ShiftType.EVENING);
        employees.add(emp5);

        Employee emp6 = new Employee("Joe");
        employees.add(emp6);

        Employee emp7 = new Employee("Terril");
        employees.add(emp7);

        Employee emp8 = new Employee("Nancy");
        employees.add(emp8);
    }
}